src/unix/lwt_io.ml

```diff
@@ -1454,6 +1454,41 @@ let with_file ?buffer ?flags ?perm ~mode filename f =
     (fun () -> f ic)
     (fun () -> close ic)
 
+let open_temp_file_with_filename:
+  ?buffer : Lwt_bytes.t ->
+  ?perm : Unix.file_perm ->
+  unit -> (string * output_channel) Lwt.t
+  = fun ?buffer ?perm ->
+    let f = Some [Unix.O_WRONLY; Unix.O_CREAT; Unix.O_TRUNC;
+                  Unix.O_NONBLOCK; Unix.O_EXCL] in
+    let rec helper n =
+      let fname = "lwt_io_temp_file_" ^ string_of_int n in
+      Lwt.catch
+        (fun () -> open_file ?buffer:buffer ?flags:f ?perm:perm ~mode:Output fname
+          >>= fun chan -> Lwt.return (fname, chan))
+        (fun exn ->
+           match exn with
+           | Unix.Unix_error _ when n < 1000 -> helper (n + 1)
+           | _ -> raise exn)
+    in
+    fun () -> helper 0
+
+let open_temp_file:
+  ?buffer : Lwt_bytes.t ->
+  ?perm : Unix.file_perm ->
+  unit -> output_channel Lwt.t
+  = fun ?buffer ?perm ->
+    let f = open_temp_file_with_filename ?buffer:buffer ?perm:perm in
+    fun () -> f () >>= (fun (_, chan) -> Lwt.return chan)
```

Comment:
Again a nit, but this would be easier to read if it didn't partially apply `open_temp_file_with_filename`, as that makes me worry about side effects if I partially apply `open_temp_file`.

How about

```ocaml
... fun ?buffer ?perm () ->
  open_temp_file_with_filename ?buffer:buffer ?perm:perm () >>= fun (_, chan) ->
  Lwt.return chan
```

or even

```ocaml
  open_temp_file_with_filename ?buffer:buffer ?perm:perm () >|= snd
```